packages/block-editor/src/components/block-list/block-wrapper.js

```diff
@@ -28,148 +28,160 @@ import { Context, SetBlockNodes } from './root-container';
+	const ref = props.ref || fallbackRef;
+	const onSelectionStart = useContext( Context );
+	const setBlockNodes = useContext( SetBlockNodes );
+	const {
+		clientId,
+		rootClientId,
+		isSelected,
+		isFirstMultiSelected,
+		isLastMultiSelected,
+		isPartOfMultiSelection,
+		enableAnimation,
+		index,
+		className,
+		name,
+		mode,
+		blockTitle,
+		wrapperProps = {},
+	} = useContext( BlockListBlockContext );
+	const {
+		initialPosition,
+		shouldFocusFirstElement,
+		isNavigationMode,
+	} = useSelect(
+		( select ) => {
+			const {
+				getSelectedBlocksInitialCaretPosition,
+				isMultiSelecting: _isMultiSelecting,
+				isNavigationMode: _isNavigationMode,
+			} = select( 'core/block-editor' );
+
+			return {
+				shouldFocusFirstElement:
+					isSelected &&
+					! _isMultiSelecting() &&
+					! _isNavigationMode(),
+				initialPosition: isSelected
+					? getSelectedBlocksInitialCaretPosition()
+					: undefined,
+				isNavigationMode: _isNavigationMode,
+			};
+		},
+		[ isSelected ]
+	);
+	const { insertDefaultBlock, removeBlock } = useDispatch(
+		'core/block-editor'
+	);
+	const [ isHovered, setHovered ] = useState( false );
+
+	// Provide the selected node, or the first and last nodes of a multi-
+	// selection, so it can be used to position the contextual block toolbar.
+	// We only provide what is necessary, and remove the nodes again when they
+	// are no longer selected.
+	useEffect( () => {
+		if ( isSelected || isFirstMultiSelected || isLastMultiSelected ) {
+			const node = ref.current;
+			setBlockNodes( ( nodes ) => ( {
+				...nodes,
+				[ clientId ]: node,
+			} ) );
+			return () => {
+				setBlockNodes( ( nodes ) => omit( nodes, clientId ) );
+			};
+		}
+	}, [ isSelected, isFirstMultiSelected, isLastMultiSelected ] );
+
+	// translators: %s: Type of block (i.e. Text, Image etc)
+	const blockLabel = sprintf( __( 'Block: %s' ), blockTitle );
+
+	// Handing the focus of the block on creation and update
+
+	/**
+	 * When a block becomes selected, transition focus to an inner tabbable.
+	 */
+	const focusTabbable = () => {
+		// Focus is captured by the wrapper node, so while focus transition
+		// should only consider tabbables within editable display, since it
+		// may be the wrapper itself or a side control which triggered the
+		// focus event, don't unnecessary transition to an inner tabbable.
+		if (
+			document.activeElement &&
+			isInsideRootBlock( ref.current, document.activeElement )
+		) {
+			return;
+		}
 
-		// Block Reordering animation
-		useMovingAnimation(
-			wrapper,
-			isSelected || isPartOfMultiSelection,
-			isSelected || isFirstMultiSelected,
-			enableAnimation,
-			index
-		);
+		// Find all tabbables within node.
+		const textInputs = focus.tabbable
+			.find( ref.current )
+			.filter( isTextField )
+			// Exclude inner blocks and block appenders
+			.filter(
+				( node ) =>
+					isInsideRootBlock( ref.current, node ) &&
+					! node.closest( '.block-list-appender' )
+			);
```

Comment:
Not strictly related to this PR, but we could optimize this by only filtering once.
```suggestion
			.filter(
				( node ) =>
					isTextField( node ) &&
					// Exclude inner blocks and block appenders.
					isInsideRootBlock( ref.current, node ) &&
					! node.closest( '.block-list-appender' )
			);
```